<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Aggregate FastQC Reports - Daniel E. Cook</title>
  <meta content='Aggregate FastQC Reports - Daniel E. Cook' property='title' />
  <meta content='Aggregate FastQC Reports - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Update: MultiQC (2019-06-21) After I originally published this script for aggregating FASTQC reports, MultiQC was published by Phil Ewels. MultiQC aggregates quality-control and other associated data from sequencing tools into an interactive report. Instead of the script below, you can simply run:
# Run this command where your *_fastqc.zip files are multiqc . This will output a repor that looks like this:
Publication
 MultiQC: Summarize analysis results for multiple tools and samples in a single report Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller Bioinformatics (2016) doi: 10." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/aggregate-fastqc-reports/" />


<meta property="article:published_time" content="2014-12-28T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2014-12-28T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.62.2" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/syntax.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/aggregate-fastqc-reports/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Aggregate FastQC Reports</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2014-12-28</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/bam' class="link silver tag">bam</a> &nbsp;&nbsp;<a href='/tags/fastq' class="link silver tag">fastq</a> &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content"><h2 id="update-multiqc-2019-06-21">Update: MultiQC (2019-06-21)</h2>
<p>After I originally published this script for aggregating FASTQC reports, <a href="https://multiqc.info">MultiQC</a> was published by <a href="https://github.com/ewels">Phil Ewels</a>. MultiQC aggregates quality-control and other associated data from sequencing tools into an interactive report. Instead of the script below, you can simply run:</p>
<div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash"><span class="c1"># Run this command where your *_fastqc.zip files are</span>
multiqc .
</code></pre></div><p>This will output a repor that looks like this:</p>
<p><img src="/multiqc.png" alt="multiqc screenshot"></p>
<p><strong>Publication</strong></p>
<blockquote>
<p><strong>MultiQC: Summarize analysis results for multiple tools and samples in a single report</strong> <!-- raw HTML omitted -->
<em>Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller</em> <!-- raw HTML omitted -->
Bioinformatics (2016) <!-- raw HTML omitted -->
doi: <a href="http://dx.doi.org/10.1093/bioinformatics/btw354">10.1093/bioinformatics/btw354</a> <!-- raw HTML omitted -->
PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/27312411">27312411</a></p>
</blockquote>
<h2 id="original-post-2014-12-28">Original Post (2014-12-28)</h2>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> is a phenomenal sequence quality assessment tool for evaluating both fastq and bam files. If you are working with a large number of sequence files (fastq), you may wish to compare results across all of them by comparing the plots that fastqc produces. I’m talking about the set of plots that look like this:</p>
<p><img src="/Uchicago-L001-CB4857_CGC-4642f-1.png" alt="fastqc"></p>
<p>FastQC can be invoked from the command line by typing <code>fastqc &lt;fastq/bam&gt;</code>, and it will produce an html report and associated zip file containing data, plots, and some ancillary files. The zip file contains an <strong>Images</strong> folder where the plots that become incorporated into the html report are stored. They are:</p>
<ul>
<li>Adapter Content</li>
<li>Duplication Levels</li>
<li>Kmer Profiles</li>
<li>Per base N Content</li>
<li>Per Base Quality</li>
<li>Per Base Sequence Content</li>
<li>Per Sequence GC Content</li>
<li>Per Sequence Quality</li>
<li>Per Tile Quality</li>
<li>Sequence Length Distribution</li>
</ul>
<p>The zipped folder also contains a file called <strong>fastqc_data.txt</strong> and <strong>summary.txt</strong>. <strong>fastqc_data.txt</strong> contains the raw data and statistics while <strong>summary.txt</strong> summarizes which tests have been passed.</p>
<p>To easily compare data across reports I wrote this short shell script (below) which will ‘aggregate’ images, statistics, and summaries by:</p>
<ol>
<li>Unzipping all the avaible fastqc zip files.</li>
<li>Creating a <strong>fq_aggregated</strong> folder, and individual folders within for each plot type.</li>
<li>Move images from each unzipped fastqc report into the folder to which it belongs, and renaming it as the filename of the report (e.g. sample name).</li>
<li>Concatenating <strong>summary.txt</strong> files as <strong>fq_aggregated</strong>/<strong>summary.txt</strong>.</li>
<li>Concatenating the basic statistics from each report into <strong>fq_aggregated</strong>/<strong>statistics.txt</strong>.</li>
</ol>
<p>Images will be reorganized as shown below:</p>
<p><img src="/aggregate_fastqc.png" alt="aggregate fastqc"></p>
<h2 id="summarytxt"><code>summary.txt</code></h2>
<p><strong>fq_aggregated</strong>/<strong>summary.txt</strong> will produce a tab delimited file that looks like this:</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>PASS</td>
<td>Basic Statistics</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  quality</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per tile  sequence  quality</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  quality scores</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per base  sequence  content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  GC  content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  N content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td></td>
<td>…</td>
<td></td>
</tr>
<tr>
<td>PASS</td>
<td>Basic Statistics</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  quality</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per tile  sequence  quality</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  quality scores</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  content</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per sequence  GC  content</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per base  N content</td>
<td>SeqB.fq</td>
</tr>
</tbody>
</table>
<h2 id="statisticstxt"><code>statistics.txt</code></h2>
<p><strong>fq_aggregated</strong>/<strong>statistics.txt</strong> will look like this:</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>PASS</td>
<td>Basic Statistics</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  quality</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per tile  sequence  quality</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  quality scores</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per base  sequence  content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  GC  content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  N content</td>
<td>SeqA.fq</td>
</tr>
<tr>
<td>d</td>
<td>…</td>
<td></td>
</tr>
<tr>
<td>PASS</td>
<td>Basic Statistics</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  quality</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per tile  sequence  quality</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per sequence  quality scores</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>PASS</td>
<td>Per base  sequence  content</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per sequence  GC  content</td>
<td>SeqB.fq</td>
</tr>
<tr>
<td>FAIL</td>
<td>Per base  N content</td>
<td>SeqB.fq</td>
</tr>
</tbody>
</table>
<h2 id="the-code">The Code</h2>
<div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash"><span class="c1"># Run this script in a directory containing zip files from fastqc. It aggregates images of each type in individual folders</span>
<span class="c1"># So looking across data is quick.</span>

<span class="nv">zips</span><span class="o">=</span><span class="sb">`</span>ls *.zip<span class="sb">`</span>

<span class="k">for</span> i in <span class="nv">$zips</span><span class="p">;</span> <span class="k">do</span>
    unzip -o <span class="nv">$i</span> <span class="p">&amp;</span>&gt;/dev/null<span class="p">;</span>
<span class="k">done</span>

<span class="nv">fastq_folders</span><span class="o">=</span><span class="si">${</span><span class="nv">zips</span><span class="p">/.zip/</span><span class="si">}</span>

rm -rf fq_aggregated <span class="c1"># Remove aggregate folder if present</span>
mkdir fq_aggregated

<span class="c1"># Rename Files within each using folder name.</span>
<span class="k">for</span> folder in <span class="nv">$fastq_folders</span><span class="p">;</span> <span class="k">do</span>
    <span class="nv">folder</span><span class="o">=</span><span class="si">${</span><span class="nv">folder</span><span class="p">%.*</span><span class="si">}</span>
    <span class="nv">img_files</span><span class="o">=</span><span class="sb">`</span>ls <span class="si">${</span><span class="nv">folder</span><span class="si">}</span>/Images/*png<span class="sb">`</span><span class="p">;</span>
    <span class="k">for</span> img in <span class="nv">$img_files</span><span class="p">;</span> <span class="k">do</span>
        <span class="nv">img_name</span><span class="o">=</span><span class="k">$(</span>basename <span class="s2">&#34;</span><span class="nv">$img</span><span class="s2">&#34;</span><span class="k">)</span><span class="p">;</span>
        <span class="nv">img_name</span><span class="o">=</span><span class="si">${</span><span class="nv">img_name</span><span class="p">%.*</span><span class="si">}</span>
        <span class="nv">new_name</span><span class="o">=</span><span class="si">${</span><span class="nv">folder</span><span class="si">}</span><span class="p">;</span>
        mkdir -p fq_aggregated/<span class="si">${</span><span class="nv">img_name</span><span class="si">}</span><span class="p">;</span>
        mv <span class="nv">$img</span> fq_aggregated/<span class="si">${</span><span class="nv">img_name</span><span class="si">}</span>/<span class="si">${</span><span class="nv">folder</span><span class="p">/_fastqc/</span><span class="si">}</span>.png<span class="p">;</span>
    <span class="k">done</span><span class="p">;</span>
<span class="k">done</span><span class="p">;</span>


<span class="c1"># Concatenate Summaries</span>
<span class="k">for</span> folder in <span class="nv">$fastq_folders</span><span class="p">;</span> <span class="k">do</span>
    <span class="nv">folder</span><span class="o">=</span><span class="si">${</span><span class="nv">folder</span><span class="p">%.*</span><span class="si">}</span>
    cat <span class="si">${</span><span class="nv">folder</span><span class="si">}</span>/summary.txt &gt;&gt; fq_aggregated/summary.txt
<span class="k">done</span><span class="p">;</span>

<span class="c1"># Concatenate Statistics</span>
<span class="k">for</span> folder in <span class="nv">$fastq_folders</span><span class="p">;</span> <span class="k">do</span>
    <span class="nv">folder</span><span class="o">=</span><span class="si">${</span><span class="nv">folder</span><span class="p">%.*</span><span class="si">}</span>
    head -n <span class="m">10</span> <span class="si">${</span><span class="nv">folder</span><span class="si">}</span>/fastqc_data.txt <span class="p">|</span> tail -n <span class="m">7</span> <span class="p">|</span> awk -v <span class="nv">f</span><span class="o">=</span><span class="si">${</span><span class="nv">folder</span><span class="p">/_fastqc/</span><span class="si">}</span> <span class="s1">&#39;{ print $0 &#34;\t&#34; f }&#39;</span> &gt;&gt; fq_aggregated/statistics.txt
    rm -rf <span class="si">${</span><span class="nv">folder</span><span class="si">}</span>
<span class="k">done</span>
</code></pre></div></div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/downgrade-a-vcf-for-viewing-in-igv-4.2-4.1/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/split-a-gff-file-into-individual-features/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>