<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Parallelize bcftools functions - Daniel E. Cook</title>
  <meta content='Parallelize bcftools functions - Daniel E. Cook' property='title' />
  <meta content='Parallelize bcftools functions - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="bcftools is a great for working with variant call files. In general, it is fast. However, I have found that the process of merging VCF files (using bcftools merge) and performing concordance checking (using bcftools gtcheck) can be a little bit slow. That is why I wrote two functions that take advantage of GNU Parallel to parallelize them.
# ~/.bashrc: executed by bash(1) for non-login shells. # see /usr/share/doc/bash/examples/startup-files (in the package bash-doc) # for examples function bam_chromosomes() { # Fetch chromosomes from a bam file samtools view -H $1 | \ grep -Po &#39;SN:(." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/parallelize-bcftools-functions/" />


<meta property="article:published_time" content="2015-11-21T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2015-11-21T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/parallelize-bcftools-functions/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home"></a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Parallelize bcftools functions</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2015-11-21</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/vcf' class="link silver tag">VCF</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content">

<p><a href="http://samtools.github.io/bcftools/">bcftools</a> is a great for working with <a href="http://www.1000genomes.org/wiki/analysis/variant%20call%20format/vcf-variant-call-format-version-41">variant call files</a>. In general, it is fast. However, I have found that the process of merging VCF files (using <code>bcftools merge</code>) and performing concordance checking (using <code>bcftools gtcheck</code>) can be a little bit slow. That is why I wrote two functions that take advantage of <a href="http://www.gnu.org/software/parallel/">GNU Parallel</a> to parallelize them.</p>

<pre><code class="language-bash"># ~/.bashrc: executed by bash(1) for non-login shells.
# see /usr/share/doc/bash/examples/startup-files (in the package bash-doc)
# for examples

function bam_chromosomes() {
    # Fetch chromosomes from a bam file
    samtools view -H $1 | \
    grep -Po 'SN:(.*)\t' | \
    cut -c 4-1000
}

function vcf_chromosomes() {
    # Fetch contigs from a vcf file.
    bcftools view -h $1 | \
    grep 'contig' | \
    egrep -o &quot;ID=([^,]+)&quot; | \
    sed 's/ID=//g' | \
    tr '\n' ' '
}


PARALLEL_CORES=6
function parallel_bcftools_merge() {
    file_set=`echo $@ | egrep -o '(\-l|\-\-file-list)(=|[ ]+)[^ ]+' | tr '=' ' ' | cut -f 2 -d ' '`
    if [ -n &quot;${file_set}&quot; ]
        then
            find_vcf=`cat ${file_set} | head -n 1`
        else
            find_vcf=`echo $@ | tr '\t' '\n' | egrep -o '[^ ]+.vcf.gz' | awk 'NR == 1 { print }' - `
    fi
    contigs=`vcf_chromosomes ${find_vcf}`
    current_dir=$(dirname ${find_vcf})
    hash_merge=`echo &quot;$@&quot; | md5sum | cut -c 1-5`
    output_prefix=&quot;${current_dir}/parallel_merge.${hash_merge}&quot;
    
    parallel --gnu --workdir ${current_dir} \
    --env args -j ${PARALLEL_CORES} \
    'bcftools merge -r {1} -O u ' $@ ' &gt; ' ${output_prefix}'.{1}.bcf' ::: ${contigs} 
    
    order=`echo $contigs | tr ' ' '\n' | awk -v &quot;prefix=${output_prefix}&quot; '{ print prefix &quot;.&quot; $0 &quot;.bcf&quot; }'`
    bcftools concat -O v ${order} | grep -v 'parallel_merge' | sed 's/##bcftools_mergeCommand=merge -r I -O u /##bcftools_mergeCommand=merge /g' | bcftools view -O u
    rm ${order}
}


PARALLEL_CORES=6
function parallel_bcftools_gtcheck() {
    find_vcf=`echo $@ | tr '\t' '\n' | egrep -o '[^ ]+.vcf.gz' | awk 'NR == 1 { print }' - `
    contigs=`vcf_chromosomes ${find_vcf}`
    current_dir=$(dirname ${find_vcf})
    hash_merge=`echo &quot;$@&quot; | md5sum | cut -c 1-5`
    output_prefix=&quot;${current_dir}/parallel_concordance.${hash_merge}&quot;
    gtcheck_options=`echo $@ | awk -v vcf=${find_vcf} '{ gsub(vcf,&quot;&quot;,$0); print $0; }'`
    parallel --gnu  -j ${PARALLEL_CORES} --workdir ${current_dir} 'bcftools view ' ${find_vcf} ' {} | \
    bcftools gtcheck ' ${gtcheck_options} ' - | \
    awk -v chrom={} &quot;/^CN/ { print \$0 \&quot;\t\&quot; chrom } \$0 !~ /.*CN.*/ { print } \$0 ~ /^# \[1\]CN/ { print \$0 \&quot;\tchrom\&quot;}&quot; - &gt; ' ${output_prefix}'.{}.tsv' ::: ${contigs}

    order=`echo $contigs | tr ' ' '\n' | awk -v &quot;prefix=${output_prefix}&quot; '{ print prefix &quot;.&quot; $0 &quot;.tsv&quot; }'`
    cat ${order} | \
    grep 'CN' | \
    awk 'NR == 1 &amp;&amp; /Discordance/ { print } NR &gt; 1 &amp;&amp; $0 !~ /Discordance/ { print }' | \
    awk '{ gsub(&quot;(# |\\[[0-9]+\\])&quot;,&quot;&quot;, $0); gsub(&quot; &quot;,&quot;_&quot;, $0); print }' | \
    cut -f 2-7 | datamash --header-in --sort --group=Sample_i,Sample_j sum Discordance  sum Number_of_sites | \
    cat &lt;(echo -e &quot;sample_i\tsample_j\tdiscordance\tnumber_of_sites&quot;) - | \
    awk 'NR == 1 { print $0 &quot;\tconcordance&quot; } NR &gt; 1 &amp;&amp; $4 == 0 { print $0 &quot;\t&quot; } NR &gt; 1 &amp;&amp; $4 &gt; 0 { print $0 &quot;\t&quot; ($4-$3)/$4 }'
    rm ${order}
}

</code></pre>

<h3 id="usage">Usage</h3>

<p>The function <code>vcf_chromosomes</code> extracts chromosomes names from a VCF file using bcftools. Parallelization occurs across chromosomes.</p>

<h3 id="parallel-bcftools-merge"><code>parallel_bcftools_merge</code></h3>

<p><code>parallel_bcftools_merge</code> is run very similar to <code>bcftools merge</code>. The only difference is that you have to pipe it into bcftools to change it to the appropriate output. For example:</p>

<pre><code class="language-bash">parallel_bcftools_merge -m all `ls *list_of_bcffiles` | bcftools view -O z &gt; merged_vcf.vcf.gz
</code></pre>

<p>The <code>parallel_bcftools_merge</code> function will generate a temporary vcf for every chromosome. You can use all flags except for <code>-O</code> with this function.</p>

<h3 id="parallel-bcftools-gtcheck"><code>parallel_bcftools_gtcheck</code></h3>

<p><code>parallel_bcftools_gtcheck</code> should not be used with <code>--all-sites</code>, or <code>--plot</code>. I recommend using this function with <code>-H</code> and <code>-G 1</code> to calculate the absolute number of differences in terms of homozygous calls between samples. Also, this function requires datamash (on OSX, install with <code>brew install datamash</code>)</p>

<p>The output file is slightly different than what bcftools normally outputs. In general, I use this function specifically to calculate conocordance between individual fastq runs - like this:</p>

<pre><code class="language-bash">parallel_bcftools_gtchek -H -G 1 union_samples.vcf.gz &gt; concordance.tsv
</code></pre>

<p>This parallelized version generates concordances for each chromosome and then merges the results together using datamash. Output looks like this:</p>

<table>
<thead>
<tr>
<th align="left">sample_i</th>
<th align="left">sample_j</th>
<th align="right">discordance</th>
<th align="right">number_of_sites</th>
<th align="right">concordance</th>
</tr>
</thead>

<tbody>
<tr>
<td align="left">BGI2-RET1-ED3049</td>
<td align="left">BGI1-RET1-ED3049</td>
<td align="right">927</td>
<td align="right">2344043</td>
<td align="right">0.999605</td>
</tr>

<tr>
<td align="left">BGI1-RET1-CB4856</td>
<td align="left">BGI1-RET1-CB4852</td>
<td align="right">144484</td>
<td align="right">2171694</td>
<td align="right">0.933469</td>
</tr>

<tr>
<td align="left">BGI1-RET1-CX11315</td>
<td align="left">BGI1-RET1-CB4852</td>
<td align="right">106964</td>
<td align="right">2721950</td>
<td align="right">0.960703</td>
</tr>

<tr>
<td align="left">BGI1-RET1-CX11315</td>
<td align="left">BGI1-RET1-CB4856</td>
<td align="right">137200</td>
<td align="right">2059983</td>
<td align="right">0.933398</td>
</tr>

<tr>
<td align="left">BGI1-RET1-DL238</td>
<td align="left">BGI1-RET1-CB4852</td>
<td align="right">148217</td>
<td align="right">2097343</td>
<td align="right">0.929331</td>
</tr>

<tr>
<td align="left">BGI1-RET1-DL238</td>
<td align="left">BGI1-RET1-CB4856</td>
<td align="right">124132</td>
<td align="right">1803664</td>
<td align="right">0.931178</td>
</tr>

<tr>
<td align="left">BGI1-RET1-DL238</td>
<td align="left">BGI1-RET1-CX11315</td>
<td align="right">146580</td>
<td align="right">1996802</td>
<td align="right">0.926593</td>
</tr>
</tbody>
</table>
</div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/an-alfred-workflow-for-generating-markdown-tables-from-your-clipboard/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/automatically-construct-/-infer-/-sense-bigquery-schema/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>