---
title: Double Checking FASTQs
author: Daniel Cook
layout: post
permalink: /double-checking-fastqs/
dsq_thread_id:
  - 2753148029
rp4wp_cached:
  - 1
rp4wp_auto_linked:
  - 1
categories:
  - Bash
  - Bioinformatics
  - Programming
  - Python
tags:
  - fastq
  - gist
  - python
---
When you have performed a sequencing project, quality control is one of the first things you will need to do. Unfortunately, [sample mix-ups][1] and other issues can and do happen. Systematic biases can also occur by machine and lane.


This script will extracting basic information from a set of [FASTQs][2] and output it to summary file (`fastq_summary.txt`). This will work with demultiplexed FASTQs generated by Illumina machines that appear in the following format:

`@HWI-EAS209_0006_FC706VJ:5:58:5894:21141#ATCACG/1`

* * *

  * **@HWI-EAS209\_0006\_FC706VJ** &#8211; Machine name
  * **5** &#8211; lane
  * **58** &#8211; tile within flowcell lane
  * **5894** &#8211; x coordinate of cluster within tile
  * **21141** &#8211; y coordinate of cluster within tile
  * **#ATCACG** &#8211; index
  * **/1** &#8211; member of pair (/1 or /2)

The script below will extract the machine name, lane, index, and pair.

{% gist 68deab00bc57a1f76ba6 %}

 [1]: http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0041815
 [2]: http://en.wikipedia.org/wiki/FASTQ_format