<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Fetch Data from UCSC Genome Browser - Daniel E. Cook</title>
  <meta content='Fetch Data from UCSC Genome Browser - Daniel E. Cook' property='title' />
  <meta content='Fetch Data from UCSC Genome Browser - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Original Post (2013-11-03) Previously, I&#8217;ve shown that you can use a mysql database browser (e.g. Sequel Pro) to access and browse the UCSC Genome Browser MySQL database.
If you have a small dataset that you would like to annotate, you can write SQL statements to fetch data. Below I show how you can use python to fetch genome coordinates by specifying gene and genome build.
# Note: Requires mysqldb; install using: # pip install MySQL-python from MySQLdb." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/fetch-data-from-ucsc-genome-browser/" />


<meta property="article:published_time" content="2013-11-03T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2013-11-03T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/syntax.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/fetch-data-from-ucsc-genome-browser/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Fetch Data from UCSC Genome Browser</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2013-11-03</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp;<a href='/tags/python' class="link silver tag">python</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content">

<h2 id="original-post-2013-11-03">Original Post (2013-11-03)</h2>

<p><a href="/accessing-the-ucsc-genome-browser-mysql-database/">Previously</a>, I&#8217;ve shown that you can use a mysql database browser (e.g. <a href="http://www.sequelpro.com/">Sequel Pro</a>) to access and browse the UCSC Genome Browser MySQL database.</p>

<p>If you have a small dataset that you would like to annotate, you can write <a href="http://www8.silversand.net/techdoc/teachsql/ch01.htm">SQL</a> statements to fetch data. Below I show how you can use python to fetch genome coordinates by specifying gene and genome build.</p>
<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="c1"># Note: Requires mysqldb; install using:</span>
<span class="c1"># pip install MySQL-python</span>
<span class="kn">from</span> <span class="nn">MySQLdb.constants</span> <span class="kn">import</span> <span class="n">FIELD_TYPE</span>
<span class="kn">import</span> <span class="nn">_mysql</span>

<span class="n">db</span> <span class="o">=</span> <span class="bp">None</span>

<span class="k">def</span> <span class="nf">fetch_gene_coordinates</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">build</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">db</span> <span class="c1"># db is global to prevent reconnecting.</span>
    <span class="k">if</span> <span class="n">db</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">print</span> <span class="s1">&#39;connect&#39;</span>
        <span class="n">conv</span><span class="o">=</span> <span class="p">{</span> <span class="n">FIELD_TYPE</span><span class="o">.</span><span class="n">LONG</span><span class="p">:</span> <span class="nb">int</span> <span class="p">}</span>
        <span class="n">db</span> <span class="o">=</span> <span class="n">_mysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">host</span><span class="o">=</span><span class="s1">&#39;genome-mysql.cse.ucsc.edu&#39;</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="s1">&#39;genome&#39;</span><span class="p">,</span> <span class="n">passwd</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">build</span><span class="p">,</span><span class="n">conv</span><span class="o">=</span><span class="n">conv</span><span class="p">)</span>
    <span class="n">db</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&#34;&#34;&#34;SELECT * FROM kgXref INNER JOIN knownGene ON kgXref.kgID=knownGene.name WHERE kgXref.geneSymbol = &#39;</span><span class="si">%s</span><span class="s2">&#39;&#34;&#34;&#34;</span> <span class="o">%</span> <span class="n">gene_name</span><span class="p">)</span>

    <span class="n">r</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">use_result</span><span class="p">()</span><span class="o">.</span><span class="n">fetch_row</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">maxrows</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">print</span> <span class="n">r</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;txStart&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;txEnd&#39;</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span>


<span class="k">print</span> <span class="n">fetch_gene_coordinates</span><span class="p">(</span><span class="s1">&#39;klf1&#39;</span><span class="p">,</span> <span class="s1">&#39;mm9&#39;</span><span class="p">)</span></code></pre></div>
<hr />

<p><strong>Note:</strong> The UCSC browser mysql resource will throttle you if you make too many queries. If you need to annotate large datasets, all of the data is freely available for download <a href="http://hgdownload-test.cse.ucsc.edu/goldenPath/">here</a>.</p>

<h2 id="update-cruzdb-2019-06-18">Update: cruzdb (2019-06-18)</h2>

<p>Since writing this post, <a href="https://github.com/brentp/cruzdb/">cruzdb</a> has been published. <code>cruzdb</code> is a python module by <a href="https://github.com/brentp">brentp</a> that greatly simplifies and facilitates queries of the UCSC genome browser.</p>

<p><a href="https://doi.org/10.1093/bioinformatics/btt534">Bioinformatics publication</a></p>

<h3 id="installation">Installation</h3>
<div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash">pip install cruzdb</code></pre></div>
<p><strong>Examples</strong></p>
<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">cruzdb</span> <span class="kn">import</span> <span class="n">Genome</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&#34;hg18&#34;</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">refGene</span><span class="o">.</span><span class="n">filter_by</span><span class="p">(</span><span class="n">name2</span><span class="o">=</span><span class="s2">&#34;MUC5B&#34;</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span>
<span class="n">refGene</span><span class="p">(</span><span class="n">chr11</span><span class="p">:</span><span class="n">MUC5B</span><span class="p">:</span><span class="mi">1200870</span><span class="o">-</span><span class="mi">1239982</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span><span class="o">.</span><span class="n">strand</span>
<span class="s1">&#39;+&#39;</span>

<span class="c1"># the first 4 introns</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span><span class="o">.</span><span class="n">introns</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
<span class="p">[(</span><span class="il">1200999L</span><span class="p">,</span> <span class="il">1203486L</span><span class="p">),</span> <span class="p">(</span><span class="il">1203543L</span><span class="p">,</span> <span class="il">1204010L</span><span class="p">),</span> <span class="p">(</span><span class="il">1204082L</span><span class="p">,</span> <span class="il">1204420L</span><span class="p">),</span> <span class="p">(</span><span class="il">1204682L</span><span class="p">,</span> <span class="il">1204836L</span><span class="p">)]</span>

<span class="c1"># the first 4 exons.</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span><span class="o">.</span><span class="n">exons</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
<span class="p">[(</span><span class="il">1200870L</span><span class="p">,</span> <span class="il">1200999L</span><span class="p">),</span> <span class="p">(</span><span class="il">1203486L</span><span class="p">,</span> <span class="il">1203543L</span><span class="p">),</span> <span class="p">(</span><span class="il">1204010L</span><span class="p">,</span> <span class="il">1204082L</span><span class="p">),</span> <span class="p">(</span><span class="il">1204420L</span><span class="p">,</span> <span class="il">1204682L</span><span class="p">)]</span>

<span class="c1"># note that some of these are not coding because they are &lt; cdsStart</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">muc5b</span><span class="o">.</span><span class="n">cdsStart</span>
<span class="il">1200929L</span></code></pre></div></div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/accessing-the-ucsc-genome-browser-mysql-database/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/export-excel-worksheets-as-individual-csvs/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>