<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Converting VCF To JSON - Daniel E. Cook</title>
  <meta content='Converting VCF To JSON - Daniel E. Cook' property='title' />
  <meta content='Converting VCF To JSON - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Recently I started developing a set of utilities called seq-collection (sc) written in nim and using the fantastic hts-nim package.
The first utility I added was a tool to convert a VCF to JSON. This tool is useful for building out an API that reads genotype data directly from the VCF format. It is possible to read specific variants or intervals of VCF files when they are indexed, allowing for fast and efficient querying of genetic data without the need for a database." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/converting-vcf-to-json/" />


<meta property="article:published_time" content="2019-09-22T01:15:53&#43;00:00"/>

<meta property="article:modified_time" content="2019-09-22T01:15:53&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/converting-vcf-to-json/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home"></a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Converting VCF To JSON</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2019-09-22</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/vcf' class="link silver tag">VCF</a> &nbsp;&nbsp;<a href='/tags/nim' class="link silver tag">nim</a> &nbsp;&nbsp;<a href='/tags/seq-collection' class="link silver tag">seq-collection</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content">

<p>Recently I started developing a set of utilities called <a href="https://github.com/danielecook/seq-collection">seq-collection</a> (<strong>sc</strong>) written in <a href="https://nim-lang.org/">nim</a> and using the fantastic <a href="https://github.com/brentp/hts-nim">hts-nim</a> package.</p>

<p>The first utility I added was a tool to convert a VCF to JSON. This tool is useful for building out an API that reads genotype data directly from the VCF format. It is possible to read specific variants or intervals of VCF files when they are indexed, allowing for fast and efficient querying of genetic data without the need for a database. Furthermore, these queries can be made over http connections making it possible to use a VCF file as a database.</p>

<p>For example, as a graduate student I developed a <a href="https://elegansvariation.org/data/browser/">genome browser</a> for <em>C. elegans</em> wild isolates. Queries are made directly on specific genomic intervals of VCF files using <code>bcftools</code>. However, a large amount of python code is used to convert VCF output ot JSON format. The <code>sc</code> utility could replace this underlying python code on the server with a simple binary and a few command line arguments to accomplish the same task. Below I have a few examples illustrating how to use <code>sc json</code>.</p>

<h1 id="installation">Installation</h1>

<p>You can download the MAC OSX binary <a href="https://github.com/danielecook/seq-collection/releases/download/0.0.1/sc_macosx">here</a>.</p>

<p>Or build from source at <a href="http://www.github.com/danielecook/seq-collection">danielecook/seq-collection</a></p>

<h1 id="usage">Usage</h1>

<pre><code class="language-bash">json

Convert a VCF to JSON

Usage:
  json [options] vcf [region ...]

Arguments:
  vcf              VCF to convert to JSON
  [region ...]     List of regions

Options:
  -i, --info=INFO            comma-delimited INFO fields; Use 'ALL' for everything
  -f, --format=FORMAT        comma-delimited FORMAT fields; Use 'ALL' for everything
  -s, --samples=SAMPLES      Set Samples (default: ALL)
  -p, --pretty               Prettify result
  -a, --array                Output as a JSON array instead of individual JSON lines
  -z, --zip                  Zip sample names with FORMAT fields (e.g. {'sample1': 25, 'sample2': 34})
  -n, --annotation           Parse ANN Fields
  --pass                     Only output variants where FILTER=PASS
  --debug                    Debug
  -h, --help                 Show this help
</code></pre>

<h1 id="examples">Examples</h1>

<h2 id="list-all-sites">List all sites</h2>

<pre><code class="language-bash">sc json tests/data/test.vcf.gz
</code></pre>

<p><strong>Output</strong></p>

<pre><code class="language-json">{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:41947,&quot;ID&quot;:&quot;.&quot;,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;T&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;]}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:105133,&quot;ID&quot;:&quot;.&quot;,&quot;REF&quot;:&quot;G&quot;,&quot;ALT&quot;:[&quot;A&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;]}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:176422,&quot;ID&quot;:&quot;.&quot;,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;G&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;]}
</code></pre>

<h2 id="pretty-output">pretty output</h2>

<p>We can &ldquo;prettify&rdquo; this output using the <code>--pretty</code> flag:</p>

<pre><code class="language-bash">sc json --pretty tests/data/test.vcf.gz
</code></pre>

<pre><code class="language-json">{
  &quot;CHROM&quot;: &quot;I&quot;,
  &quot;POS&quot;: 41947,
  &quot;ID&quot;: &quot;.&quot;,
  &quot;REF&quot;: &quot;A&quot;,
  &quot;ALT&quot;: [
    &quot;T&quot;
  ],
  &quot;QUAL&quot;: 999.0,
  &quot;FILTER&quot;: [
    &quot;PASS&quot;
  ]
}
</code></pre>

<h2 id="fetch-genotypes">Fetch Genotypes</h2>

<p>Next we can output genotype calls by specifying the <code>--FORMAT</code> flag with a <code>GT</code> argument:</p>

<pre><code class="language-bash">&gt; sc json --FORMAT=GT tests/data/test.vcf.gz
</code></pre>

<pre><code class="language-json">{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:41947,&quot;ID&quot;:null,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;T&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;GT&quot;:[[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0]]}}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:105133,&quot;ID&quot;:null,&quot;REF&quot;:&quot;G&quot;,&quot;ALT&quot;:[&quot;A&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;GT&quot;:[[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[1,1]]}}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:176422,&quot;ID&quot;:null,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;G&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;GT&quot;:[[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[1,1],[0,0],[0,0],[0,0],[0,0],[1,1],[1,1],[null,null]]}}
</code></pre>

<p>The genotypes are ordered by sample, and the numbers correspond as follows as follows:</p>

<ul>
<li><code>-1</code> → Missing</li>
<li><code>0</code> → Reference</li>
<li><code>1</code> → First ALT allele</li>
<li><code>2</code> → Second ALT allele</li>
<li><code>3</code> → etc.</li>
</ul>

<p>You can also use <code>SGT</code> to outut a string representation of genotypes (e.g. &ldquo;0/1&rdquo;). It is also possible to use <code>TGT</code> to output the actual bases:</p>

<pre><code class="language-bash"> sc json --format=TGT tests/data/test.vcf.gz
</code></pre>

<pre><code class="language-json">{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:41947,&quot;ID&quot;:null,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;T&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;TGT&quot;:[&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;]}}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:105133,&quot;ID&quot;:null,&quot;REF&quot;:&quot;G&quot;,&quot;ALT&quot;:[&quot;A&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;TGT&quot;:[&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;A/A&quot;]}}
{&quot;CHROM&quot;:&quot;I&quot;,&quot;POS&quot;:176422,&quot;ID&quot;:null,&quot;REF&quot;:&quot;A&quot;,&quot;ALT&quot;:[&quot;G&quot;],&quot;QUAL&quot;:999.0,&quot;FILTER&quot;:[&quot;PASS&quot;],&quot;FORMAT&quot;:{&quot;TGT&quot;:[&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;G/G&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;A/A&quot;,&quot;G/G&quot;,&quot;G/G&quot;,&quot;./.&quot;]}}
</code></pre>
</div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/setting-the-working-directory-in-r/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/using-gnu-parallel-for-bioinformatics/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>