<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Downloading and storing bioinformatic databases locally - Daniel E. Cook</title>
  <meta content='Downloading and storing bioinformatic databases locally - Daniel E. Cook' property='title' />
  <meta content='Downloading and storing bioinformatic databases locally - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="If you need to annotate biological data there are plenty of resources online (UCSC Genome Browser, BioMart), and plenty of programmatic tools to interact with these databases as well. But if you are going to be annotating a large dataset (like ChIP-Seq or RNA-Seq data) &#8211; you will probably not want to rely on web based services because a) It is inefficient b) You may get throttled or banned.
If you use python, it is easy to download and store data in an SQlite database." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/downloading-and-storing-bioinformatic-databases-locally/" />


<meta property="article:published_time" content="2014-01-20T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2014-01-20T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/syntax.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/downloading-and-storing-bioinformatic-databases-locally/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Downloading and storing bioinformatic databases locally</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2014-01-20</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp;<a href='/tags/python' class="link silver tag">python</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content"><p>If you need to annotate biological data there are plenty of resources online (UCSC Genome Browser, BioMart), and plenty of programmatic tools to interact with these databases as well. But if you are going to be annotating a large dataset (like ChIP-Seq or RNA-Seq data) &#8211; you will probably not want to rely on web based services because a) It is inefficient b) You may get throttled or banned.</p>

<p>If you use <strong>python</strong>, it is easy to download and store data in an SQlite database. This allows you to query the database using SQL and quickly and efficiently annotate large datasets.</p>

<p>Below you will see that is what I have done here for HapMap allele frequency data (<a href="http://hapmap.ncbi.nlm.nih.gov/downloads/genotypes/2010-08_phaseII+III/forward/">2010-08_phaseII+III</a>), and it allows me to retrieve allele frequency data from 26,278,275 rows across 11 populations instantly. The database itself is 3.22 Gb. A zipped version (~1Gb) is available <a href="https://drive.google.com/file/d/0B_6qjHtu65BDdmFBeXdGeEc2STQ/edit?usp=sharing">Here</a>.</p>

<p><a href="/Screen-Shot-2014-01-20-at-12.07.25-AM.png"><img src="/Screen-Shot-2014-01-20-at-12.07.25-AM.png" /></a></p>

<p>You will need sqlalchemy for this script to work. Install using <code>pip install sqlalchemy</code>.</p>
<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="ch">#! /usr/local/bin/Python</span>
<span class="kn">import</span> <span class="nn">sqlite3</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">sqlalchemy</span>
<span class="kn">from</span> <span class="nn">sqlalchemy</span> <span class="kn">import</span> <span class="n">Table</span><span class="p">,</span> <span class="n">Column</span><span class="p">,</span> <span class="n">Index</span><span class="p">,</span> <span class="n">Integer</span><span class="p">,</span> <span class="n">String</span><span class="p">,</span> <span class="n">Float</span><span class="p">,</span> <span class="n">MetaData</span><span class="p">,</span> <span class="n">ForeignKey</span>
<span class="kn">from</span> <span class="nn">sqlalchemy</span> <span class="kn">import</span> <span class="n">create_engine</span>
<span class="kn">import</span> <span class="nn">datetime</span>

<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>


<span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;wget -nd -r  -A &#34;allele*.gz&#34; -e robots=off &#34;http://hapmap.ncbi.nlm.nih.gov/downloads/frequencies/2010-08_phaseII+III/&#34;&#39;</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;gunzip *.gz # Unzip all the files&#39;</span><span class="p">)</span>

<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s1">&#39;hapmap.db&#39;</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;hapmap.db&#39;</span><span class="p">)</span>

<span class="n">engine</span> <span class="o">=</span> <span class="n">create_engine</span><span class="p">(</span><span class="s1">&#39;sqlite:///hapmap.db&#39;</span><span class="p">)</span>
<span class="n">conn</span> <span class="o">=</span> <span class="n">engine</span><span class="o">.</span><span class="n">connect</span><span class="p">()</span>

<span class="n">metadata</span> <span class="o">=</span> <span class="n">MetaData</span><span class="p">()</span>

<span class="n">freq</span> <span class="o">=</span> <span class="n">Table</span><span class="p">(</span><span class="s1">&#39;freq&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="n">Integer</span><span class="p">,</span> <span class="n">primary_key</span><span class="o">=</span><span class="bp">True</span><span class="p">),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;population&#39;</span><span class="p">,</span> <span class="n">String</span><span class="p">(</span><span class="mi">3</span><span class="p">)),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;rs&#39;</span><span class="p">,</span> <span class="n">Integer</span><span class="p">),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="n">String</span><span class="p">(</span><span class="mi">5</span><span class="p">)),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span> <span class="n">Integer</span><span class="p">),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;refallele&#39;</span><span class="p">,</span><span class="n">String</span><span class="p">(</span><span class="mi">3</span><span class="p">)),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;refallele_freq&#39;</span><span class="p">,</span><span class="n">Float</span><span class="p">),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;refallele_count&#39;</span><span class="p">,</span><span class="n">Integer</span><span class="p">),</span>
    <span class="c1">#</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;otherallele&#39;</span><span class="p">,</span><span class="n">String</span><span class="p">(</span><span class="mi">3</span><span class="p">)),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;otherallele_freq&#39;</span><span class="p">,</span><span class="n">Float</span><span class="p">),</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;otherallele_count&#39;</span><span class="p">,</span><span class="n">Integer</span><span class="p">),</span>
    <span class="c1">#</span>
    <span class="n">Column</span><span class="p">(</span><span class="s1">&#39;totalcount&#39;</span><span class="p">,</span><span class="n">Integer</span><span class="p">),</span>
    <span class="n">sqlite_autoincrement</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="p">)</span>



<span class="n">metadata</span><span class="o">.</span><span class="n">create_all</span><span class="p">(</span><span class="n">engine</span><span class="p">)</span>

<span class="k">for</span> <span class="n">allele_file</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&#34;allele*&#34;</span><span class="p">):</span>
    <span class="n">f</span> <span class="o">=</span> <span class="nb">file</span><span class="p">(</span><span class="n">allele_file</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="k">print</span> <span class="n">f</span>
    <span class="k">print</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
    <span class="n">pop</span> <span class="o">=</span> <span class="n">allele_file</span><span class="p">[</span><span class="n">allele_file</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="n">allele_file</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">allele_file</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="n">allele_file</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">h</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">inserts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">():</span>
        <span class="n">k</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">),</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)))</span>
        <span class="n">k</span><span class="p">[</span><span class="s1">&#39;population&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pop</span>
        <span class="n">c</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">inserts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">c</span> <span class="o">==</span> <span class="mi">1000</span><span class="p">:</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">freq</span><span class="o">.</span><span class="n">insert</span><span class="p">(),</span><span class="n">inserts</span><span class="p">)</span>
            <span class="n">inserts</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">conn</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">freq</span><span class="o">.</span><span class="n">insert</span><span class="p">(),</span><span class="n">inserts</span><span class="p">)</span>

<span class="c1"># Add indices</span>
<span class="n">Index</span><span class="p">(</span><span class="s1">&#39;population&#39;</span><span class="p">,</span> <span class="n">freq</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">engine</span><span class="p">)</span>
<span class="n">Index</span><span class="p">(</span><span class="s1">&#39;rs&#39;</span><span class="p">,</span> <span class="n">freq</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">rs</span><span class="p">)</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">engine</span><span class="p">)</span>
<span class="n">Index</span><span class="p">(</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="n">freq</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">engine</span><span class="p">)</span>
<span class="n">Index</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span> <span class="n">freq</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">pos</span><span class="p">)</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">engine</span><span class="p">)</span></code></pre></div></div>
  
        <div id='comments'></div>
  
</main>

   









<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/use-google-to-find-lecture-notes/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/alfred-workflow-for-creating-a-data-analysis-project/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  






<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}







</script>




  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });

    InstantClick.init();
    </script>
</body>
</html>