<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>From SRA Project to FASTQ - Daniel E. Cook</title>
  <meta content='From SRA Project to FASTQ - Daniel E. Cook' property='title' />
  <meta content='From SRA Project to FASTQ - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Original Post (2014-10-25) The Sequence Read Archive (SRA) contains sequence data from scientific studies stored in a special &#8216;sra&#8217; format. Data is stored in a hierarchical format:
Project ▸ Study ▸ Sample ▸ Experiment ▸ Run
Recently, I had to use the SRA to download all of the sequence data for a given project. This required querying the SRA database for all the runs in a sequencing project and converting them to FASTQs." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/from-sra-project-to-fastq/" />


<meta property="article:published_time" content="2014-10-25T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2014-10-25T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/syntax.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/from-sra-project-to-fastq/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">From SRA Project to FASTQ</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2014-10-25</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/fastq' class="link silver tag">fastq</a> &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content">

<h2 id="original-post-2014-10-25">Original Post (2014-10-25)</h2>

<p>The <a href="http://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a> contains sequence data from scientific studies stored in a special &#8216;sra&#8217; format. Data is stored in a <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/?cmd=show&amp;f=sra_sub_expl&amp;view=get_started">hierarchical format</a>:</p>

<p>Project ▸ Study ▸ Sample ▸ Experiment ▸ Run</p>

<p>Recently, I had to use the SRA to download all of the sequence data for a given project. This required querying the SRA database for all the runs in a sequencing project and converting them to FASTQs. Here&#8217;s how I did it:</p>

<p>First, you&#8217;ll need <a href="http://www.ncbi.nlm.nih.gov/books/NBK179288/">entrez direct</a>, and the sra toolkit. If you are on a mac, you can install both using <a href="homebrew.sh">homebrew</a>.</p>
<div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash">brew install edirect <span class="c1"># Entrez Direct</span>
brew install sratoolkit</code></pre></div>
<p>Once installed, the script below can be used to download all the sequence data associated with a given project. The script queries the project for all the associated sequence data, and converts to zipped FASTQs. Note that it also uses gnu parallel (to speed things up) and fastqc for quality control. These can be installed on mac using:</p>
<div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash">brew install parallel
brew install fastqc</code></pre></div><div class="highlight"><pre class="chroma"><code class="language-bash" data-lang="bash">Download_SRP_Runs<span class="o">()</span> <span class="o">{</span>
    <span class="nv">SRP_IDs</span><span class="o">=</span><span class="sb">`</span>esearch -db sra -query <span class="nv">$1</span> <span class="p">|</span> efetch -format docsum <span class="p">|</span> xtract -pattern DocumentSummary -element Run@acc <span class="p">|</span> tr <span class="s1">&#39;\t&#39;</span> <span class="s1">&#39;\n&#39;</span><span class="sb">`</span>
    <span class="k">for</span> r in <span class="si">${</span><span class="nv">SRP_IDs</span><span class="si">}</span><span class="p">;</span> <span class="k">do</span>
        <span class="nv">url</span><span class="o">=</span><span class="s2">&#34;ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/</span><span class="si">${</span><span class="nv">r</span><span class="p">:</span><span class="nv">0</span><span class="p">:</span><span class="nv">6</span><span class="si">}</span><span class="s2">/</span><span class="si">${</span><span class="nv">r</span><span class="si">}</span><span class="s2">/</span><span class="si">${</span><span class="nv">r</span><span class="si">}</span><span class="s2">.sra&#34;</span>
        wget <span class="nv">$url</span>
    <span class="k">done</span><span class="p">;</span>
<span class="o">}</span>

Download_SRP_Runs &lt;SRP ID GOES HERE&gt;

<span class="c1"># Convert to fastq</span>
parallel fastq-dump --split-files --gzip <span class="o">{}</span> ::: *.sra

<span class="c1"># Perform quality control</span>
parallel fastqc <span class="o">{}</span> ::: *.fastq.gz</code></pre></div>
<h2 id="update-sra-explorer-2019-06-20">Update: SRA Explorer (2019-06-20)</h2>

<p>The <a href="https://ewels.github.io/sra-explorer/#">SRA Explorer</a> by <a href="http://phil.ewels.co.uk/">Phil Ewels</a> can be used to generate a collection of SRA datasets and direct download links for their FASTQs.</p>
</div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/calculate-depth-and-breadth-of-coverage-from-a-bam-file/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/rename-samples-within-a-vcf/bcf/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>




  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });

    InstantClick.init();
    </script>
</body>
</html>