<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>A function for retrieving SNP data from Entrez using BioPython - Daniel E. Cook</title>
  <meta content='A function for retrieving SNP data from Entrez using BioPython - Daniel E. Cook' property='title' />
  <meta content='A function for retrieving SNP data from Entrez using BioPython - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Biopython is a great tool for interacting with biological databases. I use it to retrieve records from NCBI&#8217;s Entrez databases including Pubmed.
Unfortunately &#8211; one notable database biopython has trouble working with is the SNP database. This is due to the Bio.Entrez parser being unable to handle the XML returned from this database. One solution is to use a built in Python XML parser, but I thought I&#8217;d try to come up with an easier solution." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/a-function-for-retrieving-snp-data-from-entrez-using-biopython/" />


<meta property="article:published_time" content="2013-06-06T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2013-06-06T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/a-function-for-retrieving-snp-data-from-entrez-using-biopython/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home"></a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">A function for retrieving SNP data from Entrez using BioPython</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2013-06-06</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content"><p><a href="http://biopython.org/">Biopython</a> is a great tool for interacting with biological databases. I use it to retrieve records from <a href="http://www.ncbi.nlm.nih.gov/About/tools/restable_mol.html">NCBI&#8217;s Entrez databases</a> including Pubmed.</p>

<p>Unfortunately &#8211; one notable database biopython has trouble working with is the <a href="http://www.ncbi.nlm.nih.gov/snp">SNP</a> database. This is due to the <code>Bio.Entrez</code> parser being unable to handle the XML returned from this database. One solution is to use a built in Python XML parser, but I thought I&#8217;d try to come up with an easier solution.</p>

<p>To solve this problem &#8211; I wrote a function for retrieving SNP data, and parsing it into an array.</p>

<pre><code class="language-python">from pprint import pprint as pp
from Bio import Entrez

Entrez.email = &quot;YOUR@EMAIL.HERE&quot;

def pull_line(var_set, line):
    &quot;&quot;&quot;
    This function parses data from lines in one of three ways:
    1.) Pulls variables out of a particular line when defined as &quot;variablename=[value]&quot; - uses a string to find the variable.
    2.) Pulls variables based on a set position within a line [splits the line by '|']
    3.) Defines variables that can be identified based on a limited possible set of values - [categorical variable, specified using an array]
    &quot;&quot;&quot;
    line_set = {}
    for k,v in var_set.items():
        if type(v) == str:
            try:
                line_set[k] = [x for x in line if x.startswith(v)][0].replace(v,'')
            except:
                pass
        elif type(v) == int:
            try:
                line_set[k] = line[v]
            except:
                pass
        else:
            try:
                line_set[k] = [x for x in line if x in v][0]
            except:
                pass
    return line_set

def pull_vars(var_set,line_start,line,multi=False):
    &quot;&quot;&quot;
    Delegates and compiles data from entrez flat files dependent on whether
    the type of data trying to be pulled is contained in unique vs. non-unique lines.
    For example - the first line of the flat file is always something like this:
    rs12009 | Homo Sapiens | 9606 | etc.
    This line is unique (refers to RefSnp identifier)- and only occurs once in each flat file. On the other hand, lines
    beginning with &quot;ss[number]&quot; refer to 'submitted snp' numbers and can appear multiple times.
    &quot;&quot;&quot;
    lineset = [x.split(' | ') for x in line if x.startswith(line_start)]
    if len(lineset) == 0:
        return 
    # If the same line exists multiple times - place results into an array
    if multi == True:
        pulled_vars = []
        for line in lineset:
            # Pull date in from line and append
            pulled_vars.append(pull_line(var_set,line))
        return pulled_vars  
    else:
    # Else if the line is always unique, output single dictionary
        line = lineset[0]
        pulled_vars = {}
        return pull_line(var_set,line)

def get_snp(q):
    &quot;&quot;&quot; 
    Takes as input an array of snp identifiers and returns 
    a parsed dictionary of their data from Entrez.
    &quot;&quot;&quot;
    response = Entrez.efetch(db='SNP', id=','.join(q), rettype='flt', retmode='flt').read()
    r = {} # Return dictionary variable
    # Parse flat file response
    for snp_info in filter(None,response.split('\n\n')):
        print snp_info
        # Parse the First Line. Details of rs flat files available here:
        # ftp://ftp.ncbi.nlm.nih.gov/snp/specs/00readme.txt
        snp = snp_info.split('\n')
        # Parse the 'rs' line:
        rsId = snp[0].split(&quot; | &quot;)[0]
        r[rsId] = {}

        # rs vars
        rs_vars = {&quot;organism&quot;:1,
                   &quot;taxId&quot;:2,
                   &quot;snpClass&quot;:3,
                   &quot;genotype&quot;:&quot;genotype=&quot;,
                   &quot;rsLinkout&quot;:&quot;submitterlink=&quot;,
                   &quot;date&quot;:&quot;updated &quot;}

        # rs vars
        ss_vars = {&quot;ssId&quot;:0,
                   &quot;handle&quot;:1,
                   &quot;locSnpId&quot;:2,
                   &quot;orient&quot;:&quot;orient=&quot;,
                   &quot;exemplar&quot;:&quot;ss_pick=&quot;,
                   }

        # SNP line variables:
        SNP_vars = {&quot;observed&quot;:&quot;alleles=&quot;,
                    &quot;value&quot;:&quot;het=&quot;,
                    &quot;stdError&quot;:&quot;se(het)=&quot;,
                    &quot;validated&quot;:&quot;validated=&quot;,
                    &quot;validProbMin&quot;:&quot;min_prob=&quot;,
                    &quot;validProbMax&quot;:&quot;max_prob=&quot;,
                    &quot;validation&quot;:&quot;suspect=&quot;,
                    &quot;AlleleOrigin&quot;:['unknown',
                                    'germline',
                                    'somatic',
                                    'inherited',
                                    'paternal',
                                    'maternal',
                                    'de-novo',
                                    'bipaternal',
                                    'unipaternal',
                                    'not-tested',
                                    'tested-inconclusive'],
                    &quot;snpType&quot;:['notwithdrawn',
                               'artifact',
                               'gene-duplication',
                               'duplicate-submission',
                               'notspecified',
                               'ambiguous-location;',
                               'low-map-quality']}
        
        # CLINSIG line variables:
        CLINSIG_vars = {&quot;ClinicalSignificance&quot;:['probable-pathogenic','pathogenic','other']}

        # GMAF line variables
        GMAF_vars = {&quot;allele&quot;:&quot;allele=&quot;,
                     &quot;sampleSize&quot;:&quot;count=&quot;,
                     &quot;freq&quot;:&quot;MAF=&quot;}

        # CTG line variables
        CTG_vars = {&quot;groupLabel&quot;:&quot;assembly=&quot;,
                    &quot;chromosome&quot;:&quot;chr=&quot;,
                    &quot;physmapInt&quot;:&quot;chr-pos=&quot;,
                    &quot;asnFrom&quot;:&quot;ctg-start=&quot;,
                    &quot;asnTo&quot;:&quot;ctg-end=&quot;,
                    &quot;loctype&quot;:&quot;loctype=&quot;,
                    &quot;orient&quot;:&quot;orient=&quot;}

        # LOC line variables
        LOC_vars = {&quot;symbol&quot;:1,
                    &quot;geneId&quot;:&quot;locus_id=&quot;,
                    &quot;fxnClass&quot;:&quot;fxn-class=&quot;,
                    &quot;allele&quot;:&quot;allele=&quot;,
                    &quot;readingFrame&quot;:&quot;frame=&quot;,
                    &quot;residue&quot;:&quot;residue=&quot;,
                    &quot;aaPosition&quot;:&quot;aa_position=&quot;}

        # LOC line variables
        SEQ_vars = {&quot;gi&quot;:1,
                    &quot;source&quot;:&quot;source-db=&quot;,
                    &quot;asnFrom&quot;:&quot;seq-pos=&quot;,
                    &quot;orient&quot;:&quot;orient=&quot;}

        # Pull out variable information:
        r[rsId]['rs']       = pull_vars(rs_vars,&quot;rs&quot;,snp)
        r[rsId]['ss']       = pull_vars(ss_vars,&quot;ss&quot;,snp,True)
        r[rsId]['SNP']      = pull_vars(SNP_vars,&quot;SNP&quot;,snp)
        r[rsId]['CLINSIG']  = pull_vars(CLINSIG_vars,&quot;CLINSIG&quot;,snp)
        r[rsId]['GMAF']     = pull_vars(GMAF_vars,&quot;GMAF&quot;,snp)
        r[rsId]['CTG']      = pull_vars(CTG_vars,&quot;CTG&quot;,snp,True)
        r[rsId]['LOC']      = pull_vars(LOC_vars,&quot;LOC&quot;,snp,True)
        r[rsId]['SEQ']      = pull_vars(SEQ_vars,&quot;SEQ&quot;,snp,True)
    return r
        

snp = get_snp([&quot;12009&quot;])
print pp(snp)
</code></pre>
</div>
  
        <div id='comments'></div>
  
</main>

   









<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/create-new-file-in-finder-with-alfred-2/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/accessing-the-ucsc-genome-browser-mysql-database/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  






<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}







</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>