<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Split a GFF File into Individual Features - Daniel E. Cook</title>
  <meta content='Split a GFF File into Individual Features - Daniel E. Cook' property='title' />
  <meta content='Split a GFF File into Individual Features - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="The General Feature Format is a widely used format for annotating genome sequences. If indexed with tabix, gff files can be viewed in IGV or elsewhere. While features are organized in a nested manner (e.g. genes &gt; exons &gt; variant), you can pull out the individual types and index them, or combine only a few for viewing in your genome browser.
I was working with wormbase annotation files, which combine all the different types of features together (genes, ncRNA, mRNA, binding site, operon, G Quartets, piRNAs, etc)." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/split-a-gff-file-into-individual-features/" />


<meta property="article:published_time" content="2015-01-25T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2015-01-25T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<link href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:400,600" rel="stylesheet">
<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
}</style>
<link href='/css/tachyons.min.css' rel="stylesheet">
<link href='/css/styles.css' rel="stylesheet">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/split-a-gff-file-into-individual-features/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Split a GFF File into Individual Features</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2015-01-25</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/gff' class="link silver tag">gff</a> &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content"><p>The <a href="http://www.ensembl.org/info/website/upload/gff.html">General Feature Format</a> is a widely used format for annotating genome sequences. If indexed with tabix, gff files can be viewed in IGV or elsewhere. While features are organized in a nested manner (e.g. genes &gt; exons &gt; variant), you can pull out the individual types and index them, or combine only a few for viewing in your genome browser.</p>

<p>I was working with <a href="ftp://ftp.wormbase.org/pub/wormbase/releases/WS245/species/c_elegans/PRJNA13758/">wormbase</a> annotation files, which combine all the different types of features together (genes, ncRNA, mRNA, binding site, operon, G Quartets, piRNAs, etc). This results in a very dense track in IGV which makes it difficult to disentangle what role individual features (or features of interest) might have.</p>

<p>As a result, I wrote this very short script for splitting the individual feature types apart, sorting them, and indexing them with tabix. This way they can be selectively viewed in IGV or elsewhere.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">import</span> sys

current_feature <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;&#34;</span>

<span style="color:#66d9ef">for</span> line <span style="color:#f92672">in</span> sys<span style="color:#f92672">.</span>stdin:
    feature <span style="color:#f92672">=</span> line<span style="color:#f92672">.</span>split(<span style="color:#e6db74">&#34;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#34;</span>)[<span style="color:#ae81ff">2</span>]
    <span style="color:#66d9ef">if</span> feature <span style="color:#f92672">!=</span> current_feature:
        f <span style="color:#f92672">=</span> file(feature <span style="color:#f92672">+</span> <span style="color:#e6db74">&#34;.gff&#34;</span>, <span style="color:#e6db74">&#34;a+&#34;</span>)
    f<span style="color:#f92672">.</span>write(line)</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">gunzip -kfc &lt;GFF&gt; | grep -v ^<span style="color:#e6db74">&#34;#&#34;</span> | sort -k3,3 | python process_gff.py

<span style="color:#66d9ef">for</span> i in <span style="color:#e6db74">`</span>ls *.gff<span style="color:#e6db74">`</span>; <span style="color:#66d9ef">do</span>
    <span style="color:#f92672">(</span>grep ^<span style="color:#e6db74">&#34;#&#34;</span> $i.gff; grep -v ^<span style="color:#e6db74">&#34;#&#34;</span> $i.gff | sort -k1,1 -k4,4n<span style="color:#f92672">)</span> | bgzip &gt; $i.sorted.gff.gz;
    tabix $i.sorted.gff.gz
    rm $i.gff
<span style="color:#66d9ef">done</span></code></pre></div></div>
  
  <div class='comments'>

    <br />
    <script src="https://utteranc.es/client.js"
        repo="danielecook/danielecook.github.io"
        issue-term="pathname"
        theme="github-light"
        crossorigin="anonymous"
        async>
    </script>

</main>

   









<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/aggregate-fastqc-reports/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/an-alfred-workflow-for-working-with-sequence-data/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  






<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}







</script>




</body>
</html>