<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Parallelize by iterating over chromosomal ranges - Daniel E. Cook</title>
  <meta content='Parallelize by iterating over chromosomal ranges - Daniel E. Cook' property='title' />
  <meta content='Parallelize by iterating over chromosomal ranges - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="I have added a new utility to seq-collection called iter which generates chromosomal ranges. Lists of genomic ranges can be easily plugged into utilities such as xargs or gnu-parallel to parallelize commands.
For example:
sc iter test.bam 100,000 # Iterate on bins of 100k base pairs # Outputs &gt; I:0-999999 &gt; I:1000000-1999999 &gt; I:2000000-2999999 &gt; I:3000000-3999999 &gt; I:4000000-4999999 Note: BAMs use a 0-based coordinate system; VCFs are 1-based
This list of genomic ranges can be used to process a BAM or VCF in parallel:" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/parallelize-by-iterating-over-chromosomal-ranges/" />


<meta property="article:published_time" content="2020-01-29T01:15:53&#43;00:00"/>

<meta property="article:modified_time" content="2020-01-29T01:15:53&#43;00:00"/>








<meta name="generator" content="Hugo 0.62.2" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/parallelize-by-iterating-over-chromosomal-ranges/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home">Daniel E. Cook</a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Parallelize by iterating over chromosomal ranges</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2020-01-29</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/bam' class="link silver tag">BAM</a> &nbsp;&nbsp;<a href='/tags/vcf' class="link silver tag">VCF</a> &nbsp;&nbsp;<a href='/tags/nim' class="link silver tag">nim</a> &nbsp;&nbsp;<a href='/tags/seq-collection' class="link silver tag">seq-collection</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content"><p>I have added a new utility to <code>seq-collection</code> called <code>iter</code> which generates chromosomal ranges. Lists of genomic ranges can be easily plugged into utilities such as <code>xargs</code> or <a href="https://www.gnu.org/software/parallel/">gnu-parallel</a> to parallelize commands.</p>
<p>For example:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">sc iter test.bam 100,000 <span style="color:#75715e"># Iterate on bins of 100k base pairs</span>

<span style="color:#75715e"># Outputs</span>
&gt; I:0-999999
&gt; I:1000000-1999999
&gt; I:2000000-2999999
&gt; I:3000000-3999999
&gt; I:4000000-4999999
</code></pre></div><p><small><strong>Note:</strong> BAMs use a 0-based coordinate system; VCFs are 1-based</small></p>
<p>This list of genomic ranges can be used to process a BAM or VCF in parallel:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">
<span style="color:#66d9ef">function</span> process_chunk <span style="color:#f92672">{</span>
  <span style="color:#75715e"># Code to process chunk</span>
  vcf<span style="color:#f92672">=</span>$1
  region<span style="color:#f92672">=</span>$2
  <span style="color:#75715e"># e.g. bcftools call -m --region</span> 
  echo bcftools call --region $region $vcf <span style="color:#75715e"># ...</span>
<span style="color:#f92672">}</span>

<span style="color:#75715e"># Export the function to make it available to GNU parallel</span>
export -f process_chunk

parallel --verbose process_chunk ::: test.bam ::: <span style="color:#66d9ef">$(</span>sc iter test.bam<span style="color:#66d9ef">)</span>

</code></pre></div><p>You can also set the <code>[width]</code> option to 0 to generate a list of chromosomes.</p>
<p>See <a href="/using-gnu-parallel-for-bioinformatics/">Using GNU-Parallel for Bioinformatics</a> for a comprehensive guide on using Parallel for bioinformatics.</p>
<p><a href="https://github.com/danielecook/seq-collection">seq-collection</a> (<strong>sc</strong>) is a set of tools written in <a href="https://nim-lang.org/">nim</a> and using the fantastic <a href="https://github.com/brentp/hts-nim">hts-nim</a> package.</p>
</div>
  
        <div id='comments'></div>
  
</main>

   









<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ‚Üê <a href="https://www.danielecook.com/calculate-insert-size-metrics-faster/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  






<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}







</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>