<!doctype html>
<html lang="en">
<head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content='text/html; charset=utf-8' http-equiv='content-type' />

  <title>Calculate Depth and Breadth of Coverage From a bam File - Daniel E. Cook</title>
  <meta content='Calculate Depth and Breadth of Coverage From a bam File - Daniel E. Cook' property='title' />
  <meta content='Calculate Depth and Breadth of Coverage From a bam File - Daniel E. Cook' property='og:title' />


<meta property="og:description" content="Original Post What is the difference between depth and coverage in sequencing experiments? Actually &#8211; they refer to the same thing, the average number of reads aligned to an individual base. Previously, I had thought coverage referred to the percentage of the genome with aligned reads to it; however the more appropriate term for this is breadth of coverage. This paper more precisely defines what breadth of coverage and depth of coverage mean." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://www.danielecook.com/calculate-depth-and-breadth-of-coverage-from-a-bam-file/" />


<meta property="article:published_time" content="2014-09-20T00:00:00&#43;00:00"/>

<meta property="article:modified_time" content="2014-09-20T00:00:00&#43;00:00"/>








<meta name="generator" content="Hugo 0.55.6" />

<style type="text/css">/*https://coolors.co/afd5aa-f0f2ef-a69f98-3d3d3d-8c6057*/
:root {
  --main-color: #309fff; 
  --secondary-color: #ff622e;
  --logo-text-color: #fff;
  --body-text-color: #3d3d3d;
  --heading-text-color: #383838;
  --background-color: #fff;
  --max-width: 850px;
}</style>
<link href='https://www.danielecook.com/css/tachyons.min.min.css' integrity="" rel="stylesheet" media="screen">
<link href='https://www.danielecook.com/css/styles.min.css' integrity="" rel="stylesheet" media="screen">


<link rel="icon" 
 
  href='/favicon.ico'

type="image/x-icon"/>

<link rel="alternate" type="application/rss+xml" href="https://www.danielecook.com/calculate-depth-and-breadth-of-coverage-from-a-bam-file/feed/index.xml" title="Daniel E. Cook" />
</head>
<body class="global-font">
  <div id="nav" class="content-width center">
<nav class="flex-ns justify-between border-box pa3 pl3-l pr2-l mt1 mt0-ns" id="navbar">
  <div class="flex">
    <a class="f4 fw6 ttu no-underline dim bg-main-color pv1 ph2 br2" id="site-title" href='/' title="Home"></a>
  </div>
  
  <div class=" flex-ns mt2 mt0-ns pv1">
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/about/' title="about">about</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/' title="blog">blog</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/photos/' title="photos">photos</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/projects/' title="projects">projects</a>
    
      <a class="link dim dark-gray f6 dib mr2 mr3-l ttu tracked" href='/contact/' title="contact">contact</a>
    
  </div>
  
</nav>
</div>

  
<main class="center mv4 content-width ph3">
  <div class="f2 fw6 heading-color heading-font post-title">Calculate Depth and Breadth of Coverage From a bam File</div>
  
  <p class="silver f6 mt1 mb4 post-meta">
    <time>2014-09-20</time> 
     
    
    &nbsp;&nbsp;<a href='/tags/bam' class="link silver tag">bam</a> &nbsp;&nbsp;<a href='/tags/gist' class="link silver tag">gist</a> &nbsp;&nbsp; 
    
  </p>
  
  <div class="lh-copy post-content">

<h3 id="original-post">Original Post</h3>

<p>What is the difference between depth and coverage in sequencing experiments? Actually &#8211; <a href="https://www.biostars.org/p/6571/#6574">they refer to the same thing</a>, the average number of reads aligned to an individual base. Previously, I had thought coverage referred to the percentage of the genome with aligned reads to it; however the more appropriate term for this is <strong>breadth of coverage</strong>. <a href="http://doi.org/10.1093/bib/bbu029">This paper</a> more precisely defines what <strong>breadth of coverage</strong> and <strong>depth of coverage</strong> mean.</p>

<p><img src="/doc.png" alt="Depth vs. Breadth of Coverage" width="940" height="198" class="aligncenter size-large wp-image-805" /></p>

<p>If you need to calculate <em>depth of coverage</em> and <em>breadth of coverage</em> you can do so using the python script below. To use the script, feed the function <code>coverage</code> a bam file, and the function will return a dictionary of the depth of coverage, breadth of coverage, sum of depths (at every position), and number of bases mapped, for every contig/chromosome individually, and the entire genome as a whole.</p>

<p>Additionally, if you specify the optional second parameter specifying the mitochondrial chromosome, the script will calculate the parameters listed above for the nuclear genome and calculate the ratio of mitochondrial depth of coverage to nuclear depth of coverage. This can act as a proxy for mitochondrial count/content within a cell.</p>

<pre><code class="language-python">#
# This script calculates the depth of coverage and breadth of coverage for a given bam. 
# Outputs a dictionary containing the contig/chromosome names and the depth and breadth of coverage for each
# and for the entire genome.
#
# If you optionally specify the name of the mitochondrial chromosome (e.g. mtDNA, chrM, chrMT)
# The script will also generate breadth and depth of coverage for the nuclear genome AND the ratio
# of mtDNA:nuclearDNA; which can act as a proxy in some cases for mitochondrial count within an individual.
# 
# Author: Daniel E. Cook
# Website: Danielecook.com
#


import os
import re
from subprocess import Popen, PIPE

def get_contigs(bam):
    header, err = Popen([&quot;samtools&quot;,&quot;view&quot;,&quot;-H&quot;,bam], stdout=PIPE, stderr=PIPE).communicate()
    if err != &quot;&quot;:
        raise Exception(err)
    # Extract contigs from header and convert contigs to integers
    contigs = {}
    for x in re.findall(&quot;@SQ\WSN:(?P&lt;chrom&gt;[A-Za-z0-9_]*)\WLN:(?P&lt;length&gt;[0-9]+)&quot;, header):
        contigs[x[0]] = int(x[1])
    return contigs

def coverage(bam, mtchr = None):
    # Check to see if file exists
    if os.path.isfile(bam) == False:
        raise Exception(&quot;Bam file does not exist&quot;)
    contigs = get_contigs(bam)

    # Guess mitochondrial chromosome
    mtchr = [x for x in contigs if x.lower().find(&quot;m&quot;) == 0]
    if len(mtchr) != 1:
        mtchr = None
    else:
        mtchr = mtchr[0]

    coverage_dict = {}
    for c in contigs.keys():
        command = &quot;samtools depth -r %s %s | awk '{sum+=$3;cnt++}END{print cnt \&quot;\t\&quot; sum}'&quot; % (c, bam)
        coverage_dict[c] = {}
        coverage_dict[c][&quot;Bases Mapped&quot;], coverage_dict[c][&quot;Sum of Depths&quot;] = map(int,Popen(command, stdout=PIPE, shell = True).communicate()[0].strip().split(&quot;\t&quot;))
        coverage_dict[c][&quot;Breadth of Coverage&quot;] = coverage_dict[c][&quot;Bases Mapped&quot;] / float(contigs[c])
        coverage_dict[c][&quot;Depth of Coverage&quot;] = coverage_dict[c][&quot;Sum of Depths&quot;] / float(contigs[c])
        coverage_dict[c][&quot;Length&quot;] = int(contigs[c])

    # Calculate Genome Wide Breadth of Coverage and Depth of Coverage
    genome_length = float(sum(contigs.values()))
    coverage_dict[&quot;genome&quot;] = {}
    coverage_dict[&quot;genome&quot;][&quot;Length&quot;] = int(genome_length)
    coverage_dict[&quot;genome&quot;][&quot;Bases Mapped&quot;] = sum([x[&quot;Bases Mapped&quot;] for k, x in coverage_dict.iteritems() if k != &quot;genome&quot;])
    coverage_dict[&quot;genome&quot;][&quot;Sum of Depths&quot;] = sum([x[&quot;Sum of Depths&quot;] for k, x in coverage_dict.iteritems() if k != &quot;genome&quot;])
    coverage_dict[&quot;genome&quot;][&quot;Breadth of Coverage&quot;] = sum([x[&quot;Bases Mapped&quot;] for k, x in coverage_dict.iteritems() if k != &quot;genome&quot;]) / float(genome_length)
    coverage_dict[&quot;genome&quot;][&quot;Depth of Coverage&quot;] = sum([x[&quot;Sum of Depths&quot;] for k, x in coverage_dict.iteritems() if k != &quot;genome&quot;]) / float(genome_length)

    if mtchr != None:
        # Calculate nuclear breadth of coverage and depth of coverage
        ignore_contigs = [mtchr, &quot;genome&quot;, &quot;nuclear&quot;]
        coverage_dict[&quot;nuclear&quot;] = {}
        coverage_dict[&quot;nuclear&quot;][&quot;Length&quot;] = sum([x[&quot;Length&quot;] for k,x in coverage_dict.iteritems() if k not in ignore_contigs ])
        coverage_dict[&quot;nuclear&quot;][&quot;Bases Mapped&quot;] = sum([x[&quot;Bases Mapped&quot;] for k, x in coverage_dict.iteritems() if k not in ignore_contigs])
        coverage_dict[&quot;nuclear&quot;][&quot;Sum of Depths&quot;] = sum([x[&quot;Sum of Depths&quot;] for k, x in coverage_dict.iteritems() if k not in ignore_contigs])
        coverage_dict[&quot;nuclear&quot;][&quot;Breadth of Coverage&quot;] = sum([x[&quot;Bases Mapped&quot;] for k, x in coverage_dict.iteritems() if k not in ignore_contigs]) / float(coverage_dict[&quot;nuclear&quot;][&quot;Length&quot;])
        coverage_dict[&quot;nuclear&quot;][&quot;Depth of Coverage&quot;] = sum([x[&quot;Sum of Depths&quot;] for k, x in coverage_dict.iteritems() if k not in ignore_contigs]) / float(coverage_dict[&quot;nuclear&quot;][&quot;Length&quot;])

        # Calculate the ratio of mtDNA depth to nuclear depth
        coverage_dict[&quot;genome&quot;][&quot;mt_ratio&quot;] = coverage_dict[mtchr][&quot;Depth of Coverage&quot;] / float(coverage_dict[&quot;nuclear&quot;][&quot;Depth of Coverage&quot;])

    # Flatten Dictionary 
    coverage = []
    for k,v in coverage_dict.items():
        for x in v.items():
            coverage += [(k,x[0], x[1])]
    return coverage
</code></pre>

<p><a href="http://samtools.github.io/bcftools/">Requires BCFTools</a></p>

<h3 id="update-mosdepth-2019-06-21">Update: mosdepth (2019-06-21)</h3>

<p>If you are looking to calculate coverage, I highly recommend <a href="http://www.github.com/brentp/mosdepth">mosdepth</a>.</p>
</div>
  
        <div id='comments'></div>
  
</main>

   






<div class="tl fixed list-pages lh-copy" id="contents-list"></div>




<div class="pagination tc tr-l db fixed-l bottom-2-l right-2-l mb3 mb0-l">
  
<a id="scroll-to-top" class="f6 o-0 link br2 ph2 pv1 mb1 bg-main-color pointer" onclick="topFunction()" style="color: #fff; visibility: hidden; display: none; transition: opacity .5s, visibility .5s;" title="back to top">back to top</a>
<br>
  <p class="mb0 mt2 next_prev">
  ← <a href="https://www.danielecook.com/generate-a-bedfile-of-masked-ranges-a-fasta-file/">prev post</a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://www.danielecook.com/from-sra-project-to-fastq/">next post</a> →
  </p>
</div>



  <div class='cb'></div>
<footer class="content-width mt3 mb4 f6 center ph3 gray tc tl-l">
    <a href="http://creativecommons.org/licenses/by/3.0/deed.en_US" class='cc'>CC</a>
    <br /><br />
</footer>
  



<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.4.2/tocbot.css">
<style>.is-active-link::before { background-color: var(--secondary-color); }</style>




<script type="text/javascript">
var prevScrollpos = window.pageYOffset;
window.onscroll = function() {
  var currentScrollPos = window.pageYOffset;


  
  if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
      document.getElementById("scroll-to-top").style.display = "inline";
      document.getElementById("scroll-to-top").style.visibility = "visible";
      document.getElementById("scroll-to-top").style.opacity = "1";
  } else {
      document.getElementById("scroll-to-top").style.visibility = "hidden";
      document.getElementById("scroll-to-top").style.opacity = "0";
  }
  
  prevScrollpos = currentScrollPos;
}


function topFunction() {
  document.body.scrollTop = 0; 
  document.documentElement.scrollTop = 0; 
}






if (document.getElementById("contents-list") !== null && document.getElementsByClassName("post-content").length !== 0) { 
  tocbot.init({
    
    tocSelector: '#contents-list',
    
    contentSelector: '.post-content',
    
    headingSelector: 'h1, h2, h3',
  });
}


</script>





  <script src='https://www.danielecook.com/js/instantclick.js' integrity="" data-no-instant></script>
    <script data-no-instant>

    function utterances(index) {
        var discussion = document.getElementById('comments');
        if (!discussion) {
            return;
        }
        var script = document.createElement('script');
        script.src = 'https://utteranc.es/client.js';
        script.setAttribute('repo', 'danielecook/danielecook.com');
        script.setAttribute('issue-term', index);
        script.setAttribute('theme', 'github-light');
        script.setAttribute('crossorigin', 'anonymous');
        discussion.appendChild(script);
    }

    InstantClick.on('change', function() {
        utterances(location.pathname + location.search)
    });


    (function (i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
            m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-39176354-1', 'auto');

        InstantClick.on('change', function() {
            ga('send', 'pageview', location.pathname + location.search);
      });
       

    InstantClick.init();
    </script>
</body>
</html>